src/main/java/com/commercetools/sync/services/impl/BaseService.java

```diff
@@ -135,4 +153,156 @@
+                        + " required for resource matching.", typeName, id));
+                }
+            });
+
+        return CtpQueryUtils
+            .queryAll(syncOptions.getCtpClient(), query.get(), resourcePageConsumer)
+            .thenApply(result -> Optional.ofNullable(keyToIdCache.get(key)));
+    }
+
+    /**
+     * Given a {@link Set} of resource keys, this method fetches a set of all the resources, matching this given set of
+     * keys in the CTP project, defined in an injected {@link SphereClient}. A mapping of the key to the id
+     * of the fetched resources is persisted in an in-memory map.
+     *
+     * @param resourceKeys set of state keys to fetch matching states by
+     * @param query        a function to get query
+     * @param keyMapper    a function to get the key from the returned resource
+     * @return {@link CompletionStage}&lt;{@link Set}&lt;{@code U}&gt;&gt; in which the result of it's completion
+     *         contains a {@link Set} of all matching resources.
+     */
+    @Nonnull
+    CompletionStage<Set<U>> fetchMatchingResources(@Nonnull final Set<String> resourceKeys,
+                                                   @Nonnull final Supplier<Q> query,
+                                                   @Nonnull final Function<U, String> keyMapper) {
+        if (resourceKeys.isEmpty()) {
+            return CompletableFuture.completedFuture(Collections.emptySet());
+        }
+
+        return CtpQueryUtils
+            .queryAll(syncOptions.getCtpClient(), query.get(), Function.identity())
+            .thenApply(fetchedResources -> fetchedResources
+                .stream()
+                .flatMap(List::stream)
+                .peek(resource -> keyToIdCache.put(keyMapper.apply(resource), resource.getId()))
+                .collect(Collectors.toSet()));
+    }
+
+    /**
+     * Given a resource key, this method fetches a resource that matches this given key in the CTP project defined in a
+     * potentially injected {@link SphereClient}. If there is no matching resource an empty {@link Optional} will be
+     * returned in the returned future. A mapping of the key to the id of the fetched category is persisted in an in
+     * -memory map.
+     *
+     * @param key       the key of the resource to fetch
+     * @param query     a function to get query
+     * @param keyMapper a function to get the key from the returned resource
+     * @return {@link CompletionStage}&lt;{@link Optional}&gt; in which the result of it's completion contains an
+     *         {@link Optional} that contains the matching {@code T} if exists, otherwise empty.
+     */
+    @Nonnull
+    CompletionStage<Optional<U>> fetchResource(@Nullable final String key,
+                                               @Nonnull final Supplier<Q> query,
+                                               @Nonnull final Function<U, String> keyMapper) {
```

Comment:
I think there is no need to pass a `keyMapper` since you already pass a key as a first param. So it will be something like this:

````java
    @Nonnull
    CompletionStage<Optional<U>> fetchResource(@Nullable final String key,
                                               @Nonnull final Supplier<Q> querySupplier) {
        if (isBlank(key)) {
            return CompletableFuture.completedFuture(Optional.empty());
        }

        return syncOptions
            .getCtpClient()
            .execute(querySupplier.get())
            .thenApply(pagedQueryResult -> pagedQueryResult
                .head()
                .map(resource -> {
                    keyToIdCache.put(key, resource.getId());
                    return resource;
                }));
    }
````

and calling it would be for example for fetching a category by key:
````java
fetchResource(key, () -> CategoryQuery
            .of().plusPredicates(categoryQueryModel -> categoryQueryModel.key().is(key)));
````

Reply:
Fixed: [e4c891d](https://github.com/commercetools/commercetools-sync-java/pull/419/commits/e4c891d1d4a14e94f7bf9deb70b3ad64e2da639f)